Makefile

```diff
@@ -25,14 +25,25 @@ GIT_COMMIT := $(shell git rev-parse HEAD 2> /dev/null || true)
 
 all: binary
 
-binary:
-	go build -ldflags "-X main.gitCommit=${GIT_COMMIT}" -o ${DEST}skopeo ./cmd/skopeo
+binary: skopeo
+
+# Build a docker image (skopeobuild) that has everything we need to build.
+# Then do the build and the output (skopeo) should appear in current dir
+skopeo: cmd/skopeo
+	docker build ${DOCKER_BUILD_ARGS} -f Dockerfile.build -t skopeobuildimage .
+	docker run --rm -v ${PWD}:/src/github.com/projectatomic/skopeo \
+		skopeobuildimage make binary-local
+
+# Build w/o using Docker containers
+binary-local:
+	go build -ldflags "-X main.gitCommit=${GIT_COMMIT}" -o skopeo ./cmd/skopeo
 
 build-container:
 	docker build ${DOCKER_BUILD_ARGS} -t "$(DOCKER_IMAGE)" .
 
 clean:
 	rm -f skopeo
+	# maybe we should also erase the Docker build image too - not sure yet
```

Comment:
The canonical reference for `make` targets, to the extent anything like that exists in Linux, is https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html#Makefile-Conventions , e.g.

> `clean`: Delete all files in the current directory that are normally created by building the program. Also delete files in other directories if they are created by this makefile. However, don’t delete the files that record the configuration. …

> `distclean`: Delete all files in the current directory (or created by this makefile) that are created by configuring or building the program. If you have unpacked the source and built the program without creating any other files, ‘make distclean’ should leave only the files that were in the distribution. …

> `maintainer-clean`: Delete almost everything that can be reconstructed with this Makefile. This typically includes everything deleted by distclean, plus more: C source files produced by Bison, tags tables, Info files, and so on.  …

i.e. `make clean` is “I don’t need the build state” or even “delete the build state to guarantee a fresh rebuild”.

This model really doesn't account for build processes which create non-filesystem artifacts (and even require root privileges).

And given the Docker image cache pointed out by @duglin, removing the image name would not save any space or remove state anyway.

So I think the `clean` target is fine as it is (though without that “not sure yet” comment, please: comments in code should document design and intent of the code, or perhaps to do items (though even that is a tiny bit controversial), but if we actually decided not to do something, having a comment doubting that decision would just cause us to repeatedly and unnecessarily revisit.):

- It is clearly useful to delete the built binary and guarantee a fresh rebuild.
- The build container is fairly unrelated to the built binary and we shouldn’t need to rebuild it at all often.
- Deleting the name of the build container without affecting the cache does not affect the build process in a meaningful way.  (It _might_ be useful for `make clean` to flush the container cache so that a next rebuild truly rebuidls the build image from scratch — but honestly this is not worth automating, worrying about, and spending time on, at this point, IMHO).